bindings/java/src/main/com/apple/foundationdb/DirectBufferIterator.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * DirectBufferIterator.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb;
+
+import java.io.Closeable;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.Iterator;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A singleton that manages a pool of {@link DirectByteBuffer}, that will be shared
+ * by the {@link DirectBufferIterator} instances. It is responsibilty of user to
+ * return the borrowed buffers.
+ */
+class BufferPool {
+	static final BufferPool __instance = new BufferPool();
+	static private final int NUM_BUFFERS = 32;
+	private ArrayBlockingQueue<ByteBuffer> buffers = new ArrayBlockingQueue<>(NUM_BUFFERS);
+
+	public BufferPool() {
+		while (buffers.size() < NUM_BUFFERS) {
+			buffers.add(ByteBuffer.allocateDirect(1024 * 1024 * 4));
```

Comment:
Currently, it fails. But I was planning to fit whatever we can. I think ideally, if we use DirectBuffer, we'd probably need to give a better hint at using right size buffer. I made it big enough to just get performance numbers that we can get in good scenarios before we fallback to bad ones. 

Since DirectBuffers are expensive to allocate, we'd want to keep a pool than dynamically allocate, and use the best fit. The downside of making wrong guess and paying the cost of making another request to get rest of it will probably always be there (unless we return the old way creating results along with it too).